src/Middleware/RequestThrottling/src/RequestThrottlingEventSource.cs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.RequestThrottling
+{
+    internal sealed class RequestThrottlingEventSource : EventSource
+    {
+        public static readonly RequestThrottlingEventSource Log = new RequestThrottlingEventSource();
+        private static QueueFrame _cachedNonTimerResult
+        {
+            get
+            {
+                if (_queueFrameCache == null)
+                {
+                    _queueFrameCache = new QueueFrame { _parent = Log };
+                }
+                return _queueFrameCache.Value;
+            }
+        }
+
+        private static QueueFrame? _queueFrameCache = null;
+
+        private PollingCounter _rejectedRequestsCounter;
+        private PollingCounter _queueLengthCounter;
+        private EventCounter _queueDuration;
+
+        private long _rejectedRequests;
+        private int _queueLength;
+
+        internal RequestThrottlingEventSource()
+            : base("Microsoft.AspNetCore.RequestThrottling")
+        {
+        }
+
+        // Used for testing
+        internal RequestThrottlingEventSource(string eventSourceName)
+            : base(eventSourceName)
+        {
+        }
+
+        [Event(1, Level = EventLevel.Warning)]
+        public void RequestRejected()
+        {
+            Interlocked.Increment(ref _rejectedRequests);
+            WriteEvent(1);
+        }
+
+        [NonEvent]
+        public QueueFrame QueueTimer()
+        {
+            Interlocked.Increment(ref _queueLength);
+
+            if (IsEnabled())
+            {
+                return new QueueFrame
+                {
+                    _timer = ValueStopwatch.StartNew(),
+                    _parent = this
+                };
+            }
+
+            return _cachedNonTimerResult;
+        }
+
+        internal struct QueueFrame : IDisposable
+        {
+            internal ValueStopwatch _timer;
+            internal RequestThrottlingEventSource _parent;
+
+            public void Dispose()
+            {
+                Interlocked.Decrement(ref _parent._queueLength);
+
+                if (_parent.IsEnabled())
+                {
+                    var duration = _timer.IsActive ? _timer.GetElapsedTime().TotalMilliseconds : 0.0;
```

Comment:
In what cases wouldn't the timer be active here?

Also, if we're making QueueFrame a struct and using a ValueStopwatch, doesn't that imply we expect the request to complete without every going async and releasing the thread/stack? In that case, are milliseconds precise enough of a measurement?

Reply:
I copy-pasted this from some other tests. I'm erring on the side of caution; I do not know if inactive timers are possible.

Reply:
From benchmarking, I suspect the queues are only useful on longer running requests (5ms+). I'm open to better methods of measuring.

Reply:
> I'm erring on the side of caution; I do not know if inactive timers are possible.

As a general rule, handling cases that you know cannot or should not happen in a graceful way, ends up creating more problems than it solves.

It causes future programmers to defensively code around the same impossible case which adds verbosity and untestable code, since it's literally impossible to hit the case even in a test.

If the assumptions change, and a previously-impossible case is now possible, that usually indicates a bug. In that case you want to fail as quickly as possible. You don't want never-before-tested code to suddenly start trying to handle it gracefully.

In that case, the far more cautious thing to do is to fail fast with an Trace.Assert failure or an exception. I'm starting to really dislike Debug.Asert though.